doc/reference/security.rst

```diff
@@ -93,8 +91,233 @@ access to the field:
 
 .. todo::
 
-    field access groups apply to administrator in fields_get but not in
+    field access groups apply to the Superuser in fields_get but not in
     read/write...
 
-.. _foo: http://google.com
-.. _time module: https://docs.python.org/2/library/time.html
+.. _time module: https://docs.python.org/3/library/time.html
+
+
+.. _reference/security/pitfalls:
+
+Security Pitfalls
+=================
+
+As a developer, it is important to understand the security mechanisms and avoid
+common mistakes leading to unsecure code.
+
+Unsafe Public Methods
+---------------------
+
+Any public method can be execute via a :ref:`RPC call
+<webservices/odoo/calling_methods>` with the chosen parameters. The methods
+starting with a ``_`` are considered as private in Odoo and not callable from
+an action button or external API.
+
+On public methods, the record on which a method is executed and the parameters
+can not be trusted.
+
+.. code-block:: python
+
+    # this method is public and its arguments can not be trusted
+    def action_done(self):
+        if self.state == "draft" and self.user_has_groups('base.manager'):
+            self._set_state("done")
+
+    # this method is private and can only be called from other python method
+    def _set_state(self, new_state):
+        self.sudo().write({"state": new_state})
+
+Making a method private is obviously not enough and care must be taken to use it
+properly.
+
+Bypassing the ORM
+-----------------
+You should never use the database cursor directly when the ORM can do the same
+thing! By doing so you are bypassing all the ORM features, possibly the
+transactions, access rights and so on.
+
+And chances are that you are also making the code harder to read and probably
+less secure.
+
+.. code-block:: python
+
+    # very very wrong
+    self.env.cr.execute('SELECT id FROM auction_lots WHERE auction_id in (' + ','.join(map(str, ids))+') AND state=%s AND obj_price > 0', ('draft',))
+    auction_lots_ids = [x[0] for x in self.env.cr.fetchall()]
+
+    # no injection, but still wrong
+    self.env.cr.execute('SELECT id FROM auction_lots WHERE auction_id in %s '\
+               'AND state=%s AND obj_price > 0', (tuple(ids), 'draft',))
+    auction_lots_ids = [x[0] for x in self.env.cr.fetchall()]
+
+    # better
+    auction_lots_ids = self.search([('auction_id','in',ids), ('state','=','draft'), ('obj_price','>',0)])
```

Comment:
mebbe the code blocks could be wrapped in admonitions to outline them as dangerous / warning / good?